numbat/src/command.rs

```diff
@@ -0,0 +1,488 @@
+            code_source_id: self.code_source_id,
+        }
+    }
+
+    fn err_at_idx(&self, index: usize, err_msg: impl Into<String>) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidCommand(err_msg.into()),
+            span: self.span_from_boundary(self.inner.word_boundaries[index]),
+        }
+    }
+
+    fn err_through_end_from(&self, index: usize, err_msg: impl Into<String>) -> ParseError {
+        ParseError {
+            kind: ParseErrorKind::InvalidCommand(err_msg.into()),
+            span: self.span_through_end(index),
+        }
+    }
+
+    fn ensure_zero_args(
+        &mut self,
+        cmd: &'static str,
+        err_msg_suffix: &'static str,
+    ) -> Result<(), ParseError> {
+        let message = format!("`{}` takes 0 arguments{}", cmd, err_msg_suffix);
+        if self.inner.args.next().is_some() {
+            return Err(self.err_through_end_from(1, message));
+        }
+        Ok(())
+    }
+
+    /// Attempt to parse the input provided to Self::new as a command, such as "help",
+    /// "list <args>", "quit", etc
+    ///
+    /// Returns:
+    /// - `Ok(Command)`, if the input is a valid command with correct arguments
+    /// - `Err(ParseError)`, if the input starts with a valid command but has the wrong
+    ///   number or kind of arguments, e.g. `list foobar`
+    pub fn parse_command(&mut self) -> Result<Command, ParseError> {
+        // expect() is guaranteed to succeed; CommandParser::new ensures there is at
+        // least one word
+        let command = match &self.inner.command_kind {
+            CommandKind::Help => {
+                self.ensure_zero_args("help", "; use `info <item>` for information about an item")?;
+                Command::Help
+            }
+            CommandKind::Clear => {
+                self.ensure_zero_args("clear", "")?;
+                Command::Clear
+            }
+            CommandKind::Quit(alias) => {
+                match alias {
+                    QuitAlias::Quit => self.ensure_zero_args("quit", "")?,
+                    QuitAlias::Exit => self.ensure_zero_args("exit", "")?,
+                }
+
+                Command::Quit
+            }
+            CommandKind::Info => {
+                let err_msg = "`info` requires exactly one argument, the item to get info on";
+                let Some(item) = self.inner.args.next() else {
+                    return Err(self.err_at_idx(0, err_msg));
+                };
+
+                if self.inner.args.next().is_some() {
+                    return Err(self.err_through_end_from(1, err_msg));
+                }
+
+                Command::Info { item }
+            }
+            CommandKind::List => {
+                let items = self.inner.args.next();
+
+                if self.inner.args.next().is_some() {
+                    return Err(self.err_through_end_from(2, "`list` takes at most one argument"));
+                }
+
+                let items = match items {
```

Comment:
Any value to making this case insensitive?

Reply:
I don't know. I'd leave it case sensitive for now.